### PR TITLE
Zome type check

### PR DESCRIPTION
**Pull request by freesig**
### Summary
Small PR that adds ZomeTypeCheck and ParseEntry for checking type id's then deserializing in validation.

### TODO:
- [x] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: crates/holochain_integrity_types/src/entry.rs [hunk 1]
```diff
@@ -64,6 +65,17 @@ pub enum Entry {
...
+/// The outcome of deserializing [`Entry`] bytes to a user defined type.
+pub enum ParseEntry<T> {
+    /// The bytes successfully deserialized into `T`.
+    Valid(T),
+    /// The bytes failed to parse because of the given reason.
+    Failed(String),
+}
+
+/// The outcome of type checking and parsing a header and entry.
+pub type EntryCheck<T> = ZomeTypeCheck<ParseEntry<T>>;
```
Comment (jost-s) [inline crates/holochain_integrity_types/src/entry.rs:77]: How about something like `EntryDef::Found` and `EntryDef::NotInScope`?

Comment (freesig) [reply crates/holochain_integrity_types/src/entry.rs:77]: @jost-s `EntryDef` is already a type.

Comment (freesig) [reply crates/holochain_integrity_types/src/entry.rs:77]: But I'm not happy with the name either

Comment (maackle) [reply crates/holochain_integrity_types/src/entry.rs:77]: @freesig are these types something the zome dev is going to have to interface with?

If so what do you think about collapsing them down to one? Couldn't this just be a Result with the various reasons for not being found or parsed as the Err? Is there a benefit to nesting them?

Comment (freesig) [reply crates/holochain_integrity_types/src/entry.rs:77]: @maackle The idea is to give the user of `deserialize_from_type` more information. So they can tell if it matches the zome id  and entry def index or failed to deserialize or just hit an error calling the host.
You only want to return an error from validation if you want it to try again. Depending on the context you might want to reject if the deserialization fails or zome_id / entry def index mismatches but other times you would just ignore and pass as valid.
However I'm not happy with this PR. Maybe we can think of a better approach?

**Comment by maackle** [comment]
@freesig this is still marked draft, is it mergeable?

**Pull request closed without merging**